Plane: is_flying makeover and improved crash detect

magicrub opened the pull request:
The is_flying flag is made to be much more robust. Also added crash detection

- is_flying is populated at 5Hz instead of 1hz
- crash detection is redone from scratch taking into account all auto flights (only works in AUTO)
- added storing a small "low-resolution" history of the primary IMU for run-time statistical purposes
- added math funcions average, variance, and standard deviation
- DF log is_crashed (STAT-CRASH) and is_motionless (STAT-STILL)
- new param "CRASH_DETECT" which is a bitmask for a actions upon a crash. Only disarm is implemented here but others are easy to add such as turn on the horn or blinky emergency lights
- is_flying is not populated the first 10 seconds after a boot (may have to be re-thought for linux builds),
- much better handling of GPS loss when considering is_flying state

magicrub commented on ArduPlane/ArduPlane.cpp:
```diff
@@ -901,6 +974,109 @@ bool Plane::is_flying(void)
+        // start timer
+        crash_timer_ms = now_ms;
+
+    } else if ((now_ms >= crash_timer_ms + 2500) && !auto_state.is_crashed) {
```
should be:
(now_ms - crash_timer_ms >= 2500)

magicrub commented on ArduPlane/ArduPlane.cpp:
```diff
@@ -901,6 +974,109 @@ bool Plane::is_flying(void)
+    bool crashed_near_land_waypoint = false;
+    bool crashed = false;
+    bool been_auto_flying = (auto_state.started_flying_in_auto_ms > 0) &&
+                            (now_ms >= auto_state.started_flying_in_auto_ms + 2500);
```
should be:
(now_ms - auto_state.started_flying_in_auto_ms >= 2500)

magicrub commented on ArduPlane/ArduPlane.cpp:
```diff
@@ -845,43 +845,116 @@ void Plane::update_flight_stage(void)
+        // short drop-outs of GPS are common during flight due to banking which points the antenna in different directions
+        bool gps_lost_recently = (gps.last_fix_time_ms() > 0) && // we have locked to GPS before
+                        (gps.status() < AP_GPS::GPS_OK_FIX_2D) && // and it's lost now
+                        (now_ms < gps.last_fix_time_ms() + 5000); // but it wasn't that long ago (<5s)
```
should be:
(now_ms - gps.last_fix_time_ms() < 5000)

magicrub commented on ArduPlane/ArduPlane.cpp:
```diff
@@ -845,43 +845,116 @@ void Plane::update_flight_stage(void)
+                                    (gps.ground_speed_cm() >= ground_speed_thresh_cm);
+    bool airspeed_movement = ahrs.airspeed_estimate(&aspeed) && (aspeed >= 7);
+
+    if (now_ms < 10000) {
```
perhaps this should be larger? 20s?

magicrub commented on ArduPlane/ArduPlane.cpp:
```diff
@@ -901,6 +974,109 @@ bool Plane::is_flying(void)
     return (isFlyingProbability >= 0.9f);
 }
 
+void Plane::crash_detection_update(void)
```
function needs description

magicrub commented:
touched this up a little:
- fixed my above notes
- removed the AHRS variance test (kept variance helper functions in AP_math)
- added is_still() which utilizes the vibe stuff
- updated the mavlink heartbeat status to show EMERGENCY when is_crashed

magicrub commented:
the new stillness check is a library thing and has a new param INS_STILL_THRESH. See param desc in the commit msgs

WickedShell commented on ArduPlane/ArduPlane.cpp:
```diff
@@ -845,43 +845,117 @@ void Plane::update_flight_stage(void)
+    bool is_flying_bool;
+    uint32_t now_ms = hal.scheduler->millis();
+
+    float ground_speed_thresh_cm = (g.min_gndspeed_cm > 0) ? labs(g.min_gndspeed_cm*0.9f) : 500;
```
This reads integer values and compares against integers, so if it's an uint32_t or some such you can avoid the labs and work in pure integer math.

WickedShell commented on ArduPlane/ArduPlane.cpp:
```diff
@@ -845,43 +845,117 @@ void Plane::update_flight_stage(void)
+                    (gps_confirmed_movement || gps_lost_recently); // locked and moving, or recently lost lock
+        } else {
+            // we've never flown yet, require good GPS movement
+            is_flying_bool = airspeed_movement || // moving through the air
```
Should this be an and? Based on the comment I would have thought it should be, otherwise the comment is confusing

WickedShell commented on ArduPlane/ArduPlane.cpp:
```diff
@@ -845,43 +845,117 @@ void Plane::update_flight_stage(void)
+        auto_state.last_flying_ms = now_ms;
+
+        if ((control_mode == AUTO) &&
+            ((auto_state.started_flying_in_auto_ms == 0) || (new_is_flying != previous_is_flying)) ) {
```
It's probably a style thing, and not super relevant because compilers are clever like that but (new_is_flying != previous_is_flying) is the equivalent to (true != previous_is_flying) so just (!previous_is_flying) would work fine there.

WickedShell commented on ArduPlane/ArduPlane.cpp:
```diff
@@ -901,6 +975,120 @@ bool Plane::is_flying(void)
+                (fabsf(ahrs.roll_sensor) > 6000 || fabsf(ahrs.pitch_sensor) > 6000)) {
+                crashed = true;
+
+                // did we "crash" within 50m of the landing location? Probably just a hard landing
```
Comment is stale to actual value (75 meter) used for the check

magicrub commented on ArduPlane/ArduPlane.cpp:
```diff
@@ -845,43 +845,117 @@ void Plane::update_flight_stage(void)
+        if ((auto_state.last_flying_ms > 0) && gps_lost_recently) {
+            // we've flown before, loosen GPS constraints
+            is_flying_bool = airspeed_movement && // moving through the air
+                    (gps_confirmed_movement || gps_lost_recently); // locked and moving, or recently lost lock
```
While combing through this via @WickedShell 's review I noticed that this line gets reduced to is_flying_bool = airspeed_movement && (false || true)
It does what I want it to do: if you recently lost GPS due to a turn or some other transiet then rely on airspeed

magicrub commented:
Latest commit note:
Optimized logic by removing a forced (false || true) statement
Applied suggestions from @WickedShell regarding:
unnecessary float used where uint32 is better for ground_soeed_thresh_cm
removed 10s startup timer
updated comments

magicrub commented:
A division of any type (non base 2) is expensive. The APM 2.x is being phased out and currently has enough cpu headroom so I'm not worried about it. The Pixhawk has hardware single precision float support so a simple cast is super cheap. If it makes you feel better I could surround it with an explicit (uint32_t) case.

WickedShell commented:
I'm not bothered by the division, my only point is that there was still
floating point math happening so there was no gain in making the temp
variable a uint32_t unless the floating math goes away. (Converting is the
only expense I was really considering) That's why I originally x said it
might be a style thing.
On Aug 5, 2015 5:04 PM, "Tom Pittenger" <notifications@github.com> wrote:

> A division of any type (non base 2) is expensive. The APM 2.x is being
> phased out and currently has enough cpu headroom so I'm not worried about
> it. The Pixhawk has hardware single precision float support so a simple
> cast is super cheap. If it makes you feel better I could surround it with
> an explicit (uint32_t) case.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/diydrones/ardupilot/pull/2607#issuecomment-128189314>.
>

magicrub commented:
Well I'm later using comparing it to another uint32 so float wasn't gaining
us anything; either would work. Personally when we ditch APM 2.x support we
should nuke all "_cm" variables and make them all floats in standard
units.. but that's just me.

On Wed, Aug 5, 2015 at 5:29 PM, WickedShell <notifications@github.com>
wrote:

> I'm not bothered by the division, my only point is that there was still
> floating point math happening so there was no gain in making the temp
> variable a uint32_t unless the floating math goes away. (Converting is the
> only expense I was really considering) That's why I originally x said it
> might be a style thing.
> On Aug 5, 2015 5:04 PM, "Tom Pittenger" <notifications@github.com> wrote:
>
> > A division of any type (non base 2) is expensive. The APM 2.x is being
> > phased out and currently has enough cpu headroom so I'm not worried about
> > it. The Pixhawk has hardware single precision float support so a simple
> > cast is super cheap. If it makes you feel better I could surround it with
> > an explicit (uint32_t) case.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/diydrones/ardupilot/pull/2607#issuecomment-128189314
> >.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/diydrones/ardupilot/pull/2607#issuecomment-128193094>.
>

WickedShell commented:
Read through, looks good. Haven't flown it yet. The only outstanding thing then I'm good with it would be to base the minimum airspeed as a percentage of stall speed.

magicrub commented:
per suggestion of @WickedShell I replaced hard coded 7 m/s airspeed threshold with
    float airspeed_minimum = g.airspeed_cruise_cm * 0.005f; // = ( 0.5 * 0.01) half of cruise speed in m/s

I also combined commits so this is now ready to be pulled in by @tridge

magicrub commented:
BUMP rebased on master

magicrub commented:
I've missed the v3.4.0beta window so I'll skip the APM 2.x #ifdef stuff.

magicrub commented:
applied all changes except the AVR specific ones. Ready to merge into master (not master-AVR)

tridge commented:
applied, thanks Tom!

magicrub commented:
Thank you!
On Aug 22, 2015 5:34 PM, "Andrew Tridgell" <notifications@github.com> wrote:

> Closed #2607 <https://github.com/diydrones/ardupilot/pull/2607>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/diydrones/ardupilot/pull/2607#event-389648250>.
>
